<a href="https://colab.research.google.com/github/chlin1/Algorithms/blob/master/Elements_of_programming_interviews_in_python.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# New Section

# Primitive types

numerics(e.g. integer)
sequences (e.g. list)
mappings (e.g. dict)
classes
instances
exceptions

In [0]:
import sys
sys.maxsize
# max integer

9223372036854775807

In [0]:
2**63-1

9223372036854775807

In [0]:
sys.float_info

sys.float_info(max=1.7976931348623157e+308, max_exp=1024, max_10_exp=308, min=2.2250738585072014e-308, min_exp=-1021, min_10_exp=-307, dig=15, mant_dig=53, epsilon=2.220446049250313e-16, radix=2, rounds=1)

## **bitwise operators** [link](https://wiki.python.org/moin/BitwiseOperators)

1. x << y

Returns x with the bits shifted to the left by y places (and new bits on the right-hand-side are zeros). This is the same as multiplying x by 2\**y.
2. x >> y

Returns x with the bits shifted to the right by y places. This is the same as //'ing x by 2\**y.
3. x & y

Does a "bitwise and". Each bit of the output is 1 if the corresponding bit of x AND of y is 1, otherwise it's 0.
4. x | y

Does a "bitwise or". Each bit of the output is 0 if the corresponding bit of x AND of y is 0, otherwise it's 1.
5. ~ x

Returns the complement of x - the number you get by switching each 1 for a 0 and each 0 for a 1. This is the same as -x - 1.

6. x ^ y

Does a "bitwise exclusive or". Each bit of the output is the same as the corresponding bit in x if that bit in y is 0, and it's the complement of the bit in x if that bit in y is 1.

In [0]:
~0 # ~x =-x-1

-1

## ** 2's compliment**
Negative numbers are written with a leading one instead of a leading zero. So if you are using only 8 bits for your twos-complement numbers, then you treat patterns from "00000000" to "01111111" as the whole numbers from 0 to 127, and reserve "1xxxxxxx" for writing negative numbers. A negative number, -x, is written using the bit pattern for (x-1) with all of the bits complemented (switched from 1 to 0 or 0 to 1). So -1 is complement(1 - 1) = complement(0) = "11111111", and -10 is complement(10 - 1) = complement(9) = complement("00001001") = "11110110". This means that negative numbers go all the way down to -128 ("10000000").

use n bits to represent numbers in [-2^{n-1},2^{n-1}-1]
positive integers: [0,2^{n-1}-1] have leading bit of 0
negative integers: [-2^{n-1},-1] have leading bit of 1
(-x: first construct x, flip all bits, add 1)


any integer is stored modulo 2^n 
(x + (-x)=0 modulo 2^n)

In [0]:
# -x =~(x-1)
~(2-1)

-2

In [0]:
pow(2.71,3.14)==2.71**3.14

True

In [0]:
float('inf')
float('-inf')

-inf

In [0]:
import random


In [0]:
print(random.randrange(10))
print(random.randint(8,16))
print(random.random())
A=[3,1,5,9]
random.shuffle(A)
print(A)
random.choice(A)

2
11
0.9636769923867569
[1, 3, 9, 5]


1

##** 4.1 computing the parity of a word **

  input: a word

  output: parity of the word

1. brute force: check bit by bit 
  
  (time O(n) with n being word size)
2. use the trick: x&(x-1) = x with its lowest bit erased 

  (time O(k) with k= # of 1 in the word)
3. parity of (AB) is the same as parity of (A)^(B). iterate to the last bit and extract it by &(0b1)

  (time O(log n) with word size n)

In [0]:
def parity(x):
  # brute force time O(n)
  result = 0
  while x:
    result ^= x & 1   # x&1=1 if x=1 =0 if x=0 check the last bit is 0 or 1 # x^0=x, x^1=~x change parity
    x >>= 1  # x=x/2
  return result

In [0]:
def parity1(x):
  result = 0 
  while x:
    result ^=1
    x = x&(x-1) # drops the lowest set bit of x
  return result

In [0]:
def parity2(x):
  x ^= x>>32
  x ^= x>>16
  x ^= x>>8
  x ^= x>>4
  x ^= x>>2
  x ^= x>>1
  return x & 0b1

## **bitwise operation **

1. bin(x)

  bin(6) = '0b110'
  
  bin(6)[2:]  = '110'

2. 0xFFFF is the number FFFF in hexadecimal number system (i.e base 16 number system ) = 1+15+15\**2 +15\**3

3. 0b11 used to mask the last two digits

4. x&(x-1) clears the lowst set bit in x

5. x &~(x-1) extracts the lowest set bit of x

In [0]:
0xFFFF # 0xFFFF is the number FFFF in hexadecimal number system (i.e base 16 number system )

65535

In [0]:
15+15*16+15*16**2+15*16**3

65535

In [0]:
(2**16-1)

65535

In [0]:
0b1

1

## **4.2 swap bits **
  
  input: 64 bit integer
  
  output: swap the bits at indices i and j
  (0: least significant bit; 63 most significant bit)

1. brute force:

  use bitmasks to extract i and j bits
  
2. first test if bits to be swaped differ. if no, swap does nothing. if yes, swaping = flipping two bits

  (time O(1) independent of word size)

In [0]:
def swap_bits(x,i,j):
  ibit = x>>i & 1
  jbit = x>>j & 1
  maski = 1<<i
  maskj = 1<<j
  return (x & (~maski) & (~maskj))|(jbit<<i&maski)|(ibit<<j&maskj)

In [0]:
def swap_bits1(x,i,j):
  # extract i-th and j-th bits and see if they differ
  if (x>>i) &1 != (x>>j)&1:
    # if differ, swap = flip
    # select bits to flip with bit_mask 
    # use XOR to flip
    bit_mask = (1<<i) |(1<<j)
    x^= bit_mask
  return x

In [0]:
print(swap_bits(161,2,5))
print(swap_bits1(161,2,5))

133
133


## ** Modify a bit at a given position ** [link](https://www.geeksforgeeks.org/modify-bit-given-position/)

input: Given a number n, a position p and a binary value b 
output: we need to change the bit at position p in n to value b.

  We first create a mask that has set bit only at given position using bit wise shift.
  
      mask = 1 << position

Then to change value of bit to b, we first make it 0 using below operation

      n & ~mask

After changing it 0, we change it to b by doing or of above expression with following (b << p) & mask, i.e., we return

      (n & ~mask) | ((b << p) & mask)

In [0]:
# Python3 program to modify a bit at position 
# p in n to b. 
  
# Returns modified n. 
def modifyBit( n,  p,  b): 
    mask = 1 << p 
    return (n & ~mask) | ((b << p) & mask) 
   
# Driver code 
def main(): 
    print(modifyBit(6, 2, 0)) 
    print(modifyBit(6, 5, 1)) 
      
if __name__ == '__main__': 
    main() 
# This code is contributed by PrinciRaj1992 

2
38


## **4.3 reverse bits **

input: 64-bit unsigned integer
output: 64-bit unsigned integer with bits in reverse order

1. brute force: iterate through 32 least significant bits of input, swap each with the corresponding most significant bit

2. lookup table A[y] gives the bit-reversal of y. Given x =<y3,y2,y1,y0> with yi 16 bit integers, return <A[y0],A[y1],A[y2],A[y3]>

  (time O(n/L) with n-bit integer and L-bit cache keys)

In [0]:
def reverse_bits(x):
  mask_size = 16
  mask = 0xFFFF
  return (rev[x & mask]<<(3*mask_size) | rev[(x>>mask_size)&mask]<<(2*mask_size) \
rev[(x>>2*mask_size)&mask]<<mask_size | rev[(x>>3*mask_size)&mask])

## ** 4.4 find a closest integer with the same weight **

input: a nonnegative integer x
output: a number y, not equal x, has the same weight as x and |y-x| is as small as possible. weight of x = # of 1 in bin(x)

1. brute-force: try x-1, x+1,x-2,x+2...stop as we encounter one with the same weight as x. Poor... say x=2\**30 need to search all integers between2\**30 and 2\**29 and between 2\**30 and 2\**30 + 2\**29-1

2. y=x(flip k1-bit and flip k2-bit with k1>k2). then |x-y|=2\**k1-2\**k2 which is min if we make k1 as small as possible and k2 as close to k1

  to preserve weight-> k1 and k2 have different bit
  -> smalles k1: rightmost bit that's different from LSB
  -> k2: the very next bit
  ->swap the two rightmost sonsecutive bits that differ
  
  time (n) with integer width n
  

In [0]:
def closest_int_same_bit_count(x):
  num_bits = 64
  for i in range(num_bits-1):
    if (x>>i)&1 != (x>>(i+1))&1:
      x = x ^ (1<<i)|(1<<(i+1)) # swap bit-i and bit-(i+1)
      return x
  # raise error if all bits of x are 0 or 1
  raise ValueError('All bits are 0 or 1')

## number swap
swap two integers i and j withut using additional space

1. store i+j in i, change j to i's value and then change i to j's value

2. use ^ and x^x=0 and 0^x=x

In [0]:
def swap(i,j):
  i = i+j
  j = i-j
  i = i-j
  return i,j
  

In [0]:
def swap1(i,j):
  i = i^j
  j = j^i # j = j^(i^j) = i
  i = i^j # (i^i)^j = j 
  # show using ^ is commutative and i^i=0
  

## 4.5 compute x \* y without arithmetical operators

  input: 2 integers >0

  output: product of them

1. brute-force: repeat addition but we can't use add operator

2. result = 0 and iterate through bits of x, adding x\**k y to the result if kth bit of x is 1

  - 2\**k y = y<<k
  - addition bit-by-bit using |            
  - carry 1 if 1+1

In [0]:
def multiply(x,y):
  def add(a,b):
    running_sum, carryin, k, ta, tb = 0, 0, 1, a, b
    while ta or tb: # run through a and b. ta, tb are counter
      ak, bk = a&k, b&k # bit-k of a and b
      carryout = (ak & bk)| (ak & carryin)| (bk & carryin) # parity of carry to bit-(k+1)
      running_sum = running_sum |(ak^bk^carryin) # parity of bit-k
      carryin, k, ta, tb = (carryout<<1, k<<1, ta>>1, tb>>1) 
    return running_sum|carryin
    
  running_sum = 0
  while x: # examine each bit of x
    if x&1:
      running_sum = add(running_sum, y)
    x, y = x>>1, y<<1
  return running_sum

In [0]:
bin(13)
[bin(13&(1<<x)) for x in range(4)]

['0b1', '0b0', '0b100', '0b1000']

In [0]:
bin(13)

'0b1101'

In [0]:
multiply(10,23)

230

In [0]:
multiply(10,5)

50

## 4.6 compute x/y
  input: two positive integers
  
  output: their quotient
  
  - brute-froce: subtract y from x iteratively until the remain <y
                                                                   
  - express quotient in terms of binary
    - compute largest k s.t. 2\**k y <=x, subtract 2\**k y from x, add 2\**k to the quotient
    - in subsequent iterations, test 2\**(k-1)y, 2\**(k-2) y,... with x
    

In [0]:
def divide(x,y):
  result, power= 0, 32
  y_power = y<<power
  while x>=y:
    while y_power>x: # get max k
      y_power >>= 1
      power -= 1
      
    result += 1<<power
    x -= y_power
  return result,x,result*y+x # quotient, remainder, original x

In [0]:
divide(143,37)

(3, 32, 143)

## 4.7 compute x^y

  input: a double x and an integer y
  output: x^y
  
  - brute-force: x*x*x*x... y-1 muliplications
  
  - write y in terms of bin(y) and let y be positive
    -  x^y= x^(y0+y1\*2+y2\*2^2+..)=x^y0\* (x^2)^y1 * (x^4)^y2 *...
    - result is (x^(y/2))^2 if LSB of y is 0
    - x \* (x^(y/2))^2 if LSB of y is 1
    - if y is negative: replace x by 1/x and y by -y
    - time O(n) with n the number of bits of y

In [0]:
def power(x,y):
  result, power = 1.0, y
  if y<0:
    power, x = -power, 1.0/x
  while power:
    if power&1:
      result *= x
    power, x = power>>1, x*x
  return result

In [0]:
power(2,10)

1024.0

## 4.8 reverse digits

  input: an integer
  
  output: an integer with digits of the input in reversed order e.g. 42->24, -314->-413
  
  - brute-force: convert input to a string and traverse the string from back to front
  
  - write integer k in base 10 = k0+ k1\*10+k2\*10^2+...
  e.g. k=1132, k0=k mod 10=2, 1132/10=113, k1=113 mod 10=3, 113/10=11, k2=11 mod 10 =1, 11/10=1 k3=1
  -k<0, solve |k| and aplly the sign to the result 

In [0]:
def reverse(x):
  result, x_remaining = 0, abs(x)
  while x_remaining:
    result = result*10+x_remaining %10
    x_remaining //= 10
  return -result if x<0 else result

# Driver code 
def main(): 
    print(reverse(123456)) 
    print(reverse(-654321)) 
      
if __name__ == '__main__': 
    main() 


654321
-123456


In [0]:
1234//10

123

## 4.9 check if a decimal integer is a palindrome

  input: an integer
  
  output: true if input is a palindrome (read the same forwards and backwards) false otherwise
  
  - brute-force: convert input to a strign, pwirwise compare digits from LSD and MSD, work inwards, stop if a mismatch. time, space O(n) where n is number of digits in the input
  
  - iteratively compare the MSD and LSD, remove them from the input
    - the number of digits n of input x is n =floor(lg_10 x) +1
    - LSD=x mod 10
    - MSD = x//10^(n-1)

In [0]:
import math

def is_palindrome_number(x):
  if x<=0:
    return x==0
  
  num_digits = math.floor(math.log10(x))+1 # base 10
  mask = 10**(num_digits -1)
  for i in range(num_digits//2):
    if x // mask != x % 10:
      return False
    x %= mask # remove MSD of x
    x //= 10 # remove LSD of x
    mask /= 100 # note remove 2 digits each time
  return True

# Driver code 
def main(): 
    print(is_palindrome_number(151751)) 
    print(is_palindrome_number(157751))
    print(is_palindrome_number(-157751)) 
      
if __name__ == '__main__': 
    main() 

False
True
False


## 4.10 generate uniform random numbers

  input: two integers a\<b (zero_one_random() is provided which produces 0 or 1 with equal probability)
                          
  output: a uniform random number in [a,b]   
  
  - if [0, 2^i-1]: concatenate i bits produced by the random number generators
  
  - [a,b]->a+[0,b-a]
  
  - if b-a is not of the form 2^i-1
    - find smallest 2^i-1 > b-a
    - generate i-bit random number
    - if within [0,b-a], return it, otherwise try again until get a numberwithin the range
    - time O(log(b-a+1))
  

In [0]:
def uniform_random(lower_bound, upper_bound):
  number_of_outcomes = upper_bound - lower_bound +1
  while True:
    result, i =0, 0
    while(1<<i) < number_of_outcomes:
      # zero_one_random() is the provided random number generator
      result = (result<<1) | zero_one_random() # generate a rm 0 or 1 in LSD
      i += 1
    if result<number_of_outcomes:
      break
  return result + lower_bound

## 4.11 rectangle intersection

  input: two rectangles
  
  output: rectangle formed by their intersection
  
  
    - when the rectangles do not intersect

# Arrays

## Array boot camp

- simple brute-force solutions use O(n) space

- [3,4,5], [1]+[0]*10=[1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0], list(range(3))=[0,1,2]
- len(A), A.append(42), A.remove(4), A.insert(3,28)
- 2d array[[1,2],[3,4]]
- a in A if a is in array A
- min(A), max(A)
- for a sorted list A, bisect.bisect_left(A,6) (The returned insertion point i partitions the array a into two halves so that all(val < x for val in a[lo:i]) for the left side and all(val >= x for val in a[i:hi]) for the right side.), 
- bisect.bisect_right(A,6) or bisect.bisect(A,6)   (returns an insertion point which comes after (to the right of) any existing entries of x in a.)
- A.reverse() (inplace), reversed(A) (return an iterator)
- A.sort() (inplace), sorte(A) (return a copy)
- del A[i] (delete ith element), del A[i,j] (remove the slice)
- A[i,j,k] slice between i and j(exclusive) with span k
- A[::-1] (reverse list)
- A[k:]+A[:k] rotates A by k to the left
- B=A[;] does a shallow copy of A into B
- [x\**2 for x in range(6)], [x\**2 for x in range(6) if x%2==0]
-[(x,y) for x in A for y in B]
- M 2d list->1d list [x for row in M for x in row]



In [0]:
list(range(3))

[0, 1, 2]

In [0]:
a=[[1,2],[3,4]]


In [0]:
b=a
b


[[1, 2], [3, 4]]

In [0]:
b=a[:]
b

[[1, 2], [3, 4]]

In [0]:
M=[[10,20],[30,40]]
[x for row in M for x in row]
# [x for x in row for row in M] (error: name 'row' is not defined)

[10, 20, 30, 40]

In [0]:
import bisect
data = [('red', 5), ('blue', 1), ('yellow', 8), ('black', 0)]
data.sort(key=lambda r: r[1])
print(data)
keys = [r[1] for r in data]
print(keys)
print(data[bisect.bisect_left(keys, 0)]) # left to 0 so return 0
print(data[bisect.bisect_left(keys, 1)]) # left to 1 so return 1
print(data[bisect.bisect_right(keys, 1)]) # right to 1 so return 2
print(data[bisect.bisect(keys, 1)]) # same as bisect_right

[('black', 0), ('blue', 1), ('red', 5), ('yellow', 8)]
[0, 1, 5, 8]
('black', 0)


('red', 5)

The bisect() function can be useful for numeric table lookups. This example uses bisect() to look up a letter grade for an exam score (say) based on a set of ordered numeric breakpoints: 90 and up is an ‘A’, 80 to 89 is a ‘B’, and so on:

In [0]:
def grade(score, breakpoints=[60, 70, 80, 90], grades='FDCBA'):
        i = bisect.bisect(breakpoints, score)
        return grades[i]
    
[grade(score) for score in [33, 99, 77, 70, 89, 90, 100]]

['F', 'A', 'C', 'C', 'B', 'A', 'A']

## 5.1 the Dutch national flag problem

  input : an array A and an index i
  output: all elements<A[i]+elements=A[i]+elements>A[i]
  
  - brute-force: form three lists which less than, equal to and greater than the pivot. time O(n) and space O(n) where n = len(A)
  
  - space O(1) time O(n^2): first, iterate through A from index 0, move all elements < pivot to the start of A. second, move all elements > pivot to the end of A
  
  - space O(1) time O(n): single pass moves all element < pivot to the beginning. second pass moves larger elements to the end. use smaller and larger indices and iterate through A
  
  - space O(1), time O(n): maintain 4 subarrays: bottom(<pivot), middle(=pivot), unclassified and top(>pivot). Iterate through unclassified, move elements into one group according to the order between the incoming unclassified element and the pivot
                                                              

In [0]:
import sys
import random

def dutch_flag_partition(pivot_index, A):
  pivot = A[pivot_index]
  # First pass: group elements smaller than pivot
  for i in range(len(A)):
    # look for a smaller element
    for j in range(i+1,len(A)):
      if A[j] < pivot:
          A[i], A[j] = A[j], A[i]
          break
  # Second pass: group elements larger than pivot
  for i in reversed(range(len(A))):
    if A[i] < pivot:
      break
      # look for a larger element. stop when we reach an element <pivot since 
      # first pass has moved them to the start of A
      for j in reversed(range(i)):
        if A[j]>pivot:
          A[i], A[j] = A[j], A[i]
          break
          
# Driver code 
def main(): 
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 15)
    A = [random.randint(0, 3) for _ in range(n)]
    pivot_index = random.randrange(n)
    print(A,pivot_index)
    dutch_flag_partition(pivot_index, A)
    print(A)
  


if __name__ == '__main__':
    main()

[3, 3, 3, 2, 1, 3, 2, 0, 2, 0] 6
[1, 0, 0, 2, 3, 3, 2, 3, 2, 3]


In [0]:
def detch_flag_partition(pivot_index, A):
  pivot = A[pivot_index]
  # first pass: group elements smaller than pivot
  smaller = 0
  for i in range(len(A)):
    if A[i]<pivot:
      A[i],A[smaller]=A[smaller],A[i]
      smaller +=1
  # second pass: group elements larger than pivot
  larger = len(A)-1
  for i in reversed(range(len(A))):
    if A[i]<pivot:
      break
    if A[i] >pivot:
      A[i], A[larger]=A[larger],A[i]
      larger -= 1
    
# Driver code 
def main(): 
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 15)
    A = [random.randint(0, 3) for _ in range(n)]
    pivot_index = random.randrange(n)
    print(A,pivot_index)
    dutch_flag_partition(pivot_index, A)
    print(A)
  


if __name__ == '__main__':
    main()

[1, 2, 0, 3, 2, 1, 3] 3
[2, 0, 2, 1, 1, 3, 3]


In [0]:
def dutch_flag_partition(pivot_index, A):
  pivot = A[pivot_index]
  # keep the following invariants during partitioning
  # bottom group: A[:smaller]
  # middle group: A[smaller:equall]
  # unclassified group: A[equal:larger]
  # top group: A[larger:]
  smaller, equal, larger= 0,0,len(A)
  # keep iterating as long as there is an unclassified element
  while equal < larger:
    # A[equal] is the incoming unclassified element
    if A[equal]<pivot:
      A[smaller],A[equal]=A[equal],A[smaller]
      smaller, equal = smaller +1, equal +1
    elif A[equal]==pivot:
      equal += 1
    else: # A[equal]>pivot
      A[equal],A[larger]=A[larger],A[equal]
      larger -= 1
      
# Driver code 
def main(): 
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 15)
    A = [random.randint(0, 3) for _ in range(n)]
    pivot_index = random.randrange(n)
    print(A,pivot_index)
    dutch_flag_partition(pivot_index, A)
    print(A)
  


if __name__ == '__main__':
    main()

[0, 0, 3, 0, 3, 0, 3, 2, 3, 3, 3] 10
[0, 0, 0, 0, 2, 3, 3, 3, 3, 3, 3]


In [0]:
for i in range(10):
  for j in range(i+1, 10):
    if j<4:
      print([i,j])
      break

[0, 1]
[1, 2]
[2, 3]


In [0]:
for i in range (5):
  if i==3:
    break
  print(i)

0
1
2


In [0]:
for i in range (5):
  if i==3:
    continue
  print(i)

0
1
2
4


In [0]:
for i in reversed(range(3)):
  print(i)

2
1
0


In [0]:
import random 
random.sample(range(1, 100), 3)


[28, 90, 94]

## random

- random.randint(a, b)
Return a random integer N such that a <= N <= b.
- random.choice(seq)
Return a random element from the non-empty sequence seq. If seq is empty, raises IndexError.
- random.random()
Return the next random floating point number in the range [0.0, 1.0).
- random.choice(seq)
Return a random element from the non-empty sequence seq. If seq is empty, raises IndexError.
- 

In [0]:
import random
random.random()        # Random float x, 0.0 <= x < 1.0
#0.37444887175646646
random.uniform(1, 10)  # Random float x, 1.0 <= x < 10.0
#1.1800146073117523
random.randint(1, 10)  # Integer from 1 to 10, endpoints included
#7
random.randrange(0, 101, 2)  # Even integer from 0 to 100
#26
random.choice('abcdefghij')  # Choose a random element
#'c'

items = [1, 2, 3, 4, 5, 6, 7]
random.shuffle(items)
items
#[7, 3, 2, 5, 6, 4, 1]

random.sample([1, 2, 3, 4, 5],  3)  # Choose 3 elements
#[4, 1, 5]

## 5.2 increment an arbitrary-precision integer

  input: an arry of digits encoding a nonnegative decimal integer D
  
  output: an array representing the integer D+1
  e.g. input = <1,2,9>, output=<1,3,0>
  
  - brute-force:covert the array into integer and then convert the result back to an array of digits. fail when input encodes integers outside the range of integer type (overflow)
  
  - add digits from the LSD and propagate carries
    - time O(n) where n is len(A)
  

In [0]:
def plus_one(A):
  A[-1] += 1
  for i in reversed(range(1,len(A))):
    if A[i]!=10:
      break
    A[i]=0
    A[i-1]+=1
    
  if A[0]==10:
    # there is a carry-out, so we need one more digit to store the result
    # append a 0 at the end of the array and update the first entry to 1
    A[0]=1
    A.append(0)
  return A



def rand_vector(length):
    if not length:
        return [0]
    return [random.randint(1, 9)] + [random.randint(0, 9)
                                     for _ in range(length - 1)]

def small_test():
    result = plus_one([9, 9])
    assert result == [1, 0, 0]
    result = plus_one([3, 1, 4])
    assert result == [3, 1, 5]


def main():
    small_test()
    n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(0, 1000)
    A = rand_vector(n)
    print(*A, sep='')
    result = plus_one(A)
    print(*result, sep='')
    print(*result,sep=',')
    
if __name__ == '__main__':
    main()

8221815606897321454554455947702007899227617596642578972164186274152620572424894022091833171000345533307514576746099792989552712748680033753029105853480055724265949305342292204227483514128151144377667787750038303099138775966475635021615763892560814866651496012673028249768627126892440628859421879316834556184636645615078687954189679694162828892661331576001158688830757671597217951465045233317988734651071779049179250224759146222229763271760893791854966691295123798398152688989381781751656700237143530881339
82218156068973214545544559477020078992276175966425789721641862741526205724248940220918331710003455333075145767460997929895527127486800337530291058534800557242659493053422922042274835141281511443776677877500383030991387759664756350216157638925608148666514960126730282497686271268924406288594218793168345561846366456150786879541896796941628288926613315760011586888307576715972179514650452333179887346510717790491792502247591462222297632717608937918549666912951237983981526889893817817516567002371

## 5.3 multiply two arbitrary-prevision integer

  input: 2 arrays which represent two integers with one digit per array entry e.g. <-7,1,3>=-713
  
  output: an array representing their product
  
    - time O(nm): number of digits for the product is at most n+m for n and m diit operands

In [0]:
def multiply(num1, num2):
  sign = -1 if (num1[0]<0) ^ (num2[0]<0) else 1
  num1[0],num2[0]=abs(num1[0]),abs(num2[0])
  
  result = [0] *(len(num1)+len(num2))
  for i in reversed(range(len(num1))):
    for j in reversed(range(len(num2))):
      result[i+j+1] += num1[i]*num2[j]
      result[i+j] += result[i+j+1]//10
      result[i+j+1] %= 10
  # remove the leading zeroes
  result = result[next((i for i,x in enumerate(result) if x !=0), len(result)):] or [0]
  return [sign* result[0]] + result[1:]



def simple_test():
    assert multiply([0], [-1, 0, 0, 0]) == [0]
    assert multiply([0], [1, 0, 0, 0]) == [0]
    assert multiply([9], [9]) == [8, 1]
    assert multiply([9], [9, 9, 9, 9]) == [8, 9, 9, 9, 1]
    assert multiply([1, 3, 1, 4, 1, 2],[-1, 3, 1, 3, 3, 3, 2])==[-1, 7, 2, 5, 8, 7, 5, 8, 4, 7, 8, 4]
    assert multiply([7, 3], [-3]) == [-2, 1, 9]


def main():
    simple_test()
  
if __name__ == '__main__':
    main()

## next
[next(iterator, default)](https://www.programiz.com/python-programming/methods/built-in/next)

  - iterator - next() retrieves next item from the iterator

  - default (optional) - this value is returned if the iterator is exhausted (no items left)

In [0]:
random = [5, 9]

# converting list to iterator
randomIterator = iter(random)

# Output: 5
print(next(randomIterator, '-1'))

# Output: 9
print(next(randomIterator, '-1'))

# randomIterator is exhausted
# Output: '-1'
print(next(randomIterator, '-1'))
print(next(randomIterator, '-1'))
print(next(randomIterator, '-1'))

5
9
-1
-1
-1


In [0]:
a=(i for i,x in enumerate([0,0,0,1,2,]) if x !=0)
print(list(a)) # run over the iterator to get list so next will give default value
print(next(a,-1)) 
print(next(a,-1))

[3, 4]
-1
-1


In [0]:
a=(i for i,x in enumerate([0,0,0,1,2,]) if x !=0)
print(next(a,-1)) 
print(next(a,-1))

3
4


## [copy in Python (Deep Copy and Shallow Copy)](https://www.geeksforgeeks.org/copy-python-deep-copy-shallow-copy/)

  - Deep copy is a process in which the copying process occurs recursively. It means first constructing a new collection object and then recursively populating it with copies of the child objects found in the original. In case of deep copy, a copy of object is copied in other object. It means that any changes made to a copy of object do not reflect in the original object. In python, this is implemented using “deepcopy()” function.
  
  - A shallow copy means constructing a new collection object and then populating it with references to the child objects found in the original. The copying process does not recurse and therefore won’t create copies of the child objects themselves. In case of shallow copy, a reference of object is copied in other object. It means that any changes made to a copy of object do reflect in the original object. In python, this is implemented using “copy()” function.
  
 - The difference between shallow and deep copying is only relevant for compound objects (objects that contain other objects, like lists or class instances):

    - A shallow copy constructs a new compound object and then (to the extent possible) inserts references into it to the objects found in the original.
    - A deep copy constructs a new compound object and then, recursively, inserts copies into it of the objects found in the original.

In [0]:
# Python code to demonstrate copy operations 
  
# importing "copy" for copy operations 
import copy 
  
# initializing list 1 
li1 = [1, 2, [3,5], 4] 
  
# using deepcopy to deep copy  
li2 = copy.deepcopy(li1) 
  
# original elements of list 
print ("The original elements before deep copying") 
for i in range(0,len(li1)): 
    print (li1[i],end=" ") 
  
print("\r") 
  
# adding and element to new list 
li2[2][0] = 7
  
# Change is reflected in l2  
print ("The new list of elements after deep copying ") 
for i in range(0,len( li1)): 
    print (li2[i],end=" ") 
  
print("\r") 
  
# Change is NOT reflected in original list 
# as it is a deep copy 
print ("The original elements after deep copying") 
for i in range(0,len( li1)): 
    print (li1[i],end=" ") 

The original elements before deep copying
1 2 [3, 5] 4 
The new list of elements after deep copying 
1 2 [7, 5] 4 
The original elements after deep copying
1 2 [3, 5] 4 

In [0]:
# Python code to demonstrate copy operations 
  
# importing "copy" for copy operations 
import copy 
  
# initializing list 1 
li1 = [1, 2, [3,5], 4] 
  
# using copy to shallow copy  
li2 = copy.copy(li1) 
  
# original elements of list 
print ("The original elements before shallow copying") 
for i in range(0,len(li1)): 
    print (li1[i],end=" ") 
  
print("\r") 
  
# adding and element to new list 
li2[2][0] = 7
  
# checking if change is reflected 
print ("The original elements after shallow copying") 
for i in range(0,len( li1)): 
    print (li1[i],end=" ") 

The original elements before shallow copying
1 2 [3, 5] 4 
The original elements after shallow copying
1 2 [7, 5] 4 

## 5.4 Advancing through an array

  input: an array of n integers where A[i] denotes the max you can advance from index i
  
  output: True if it is possible to advance to the last index from index 0
  
  - iterate through A, track the furthest index we can advance to, the furthest index from index i is i+A[i]
  - time O(n), space O(1)
  
  

In [0]:
def can_reach_end(A):
  furthest_reach_so_far, last_index = 0 ,len(A)-1
  i=0
  while i <=furthest_reach_so_far and furthest_reach_so_far < last_index:
    furthest_reach_so_far = max(furthest_reach_so_far, A[i]+i)
    i += 1
  return furthest_reach_so_far >= last_index


print(can_reach_end([3,3,1,0,2,0,1]))
print(can_reach_end([3,2,0,0,2,0,1]))

  

True
False


## 5.5 delete duplicates from a sorted array

  input: an sorted array
  
  output: all duplicated are removed and the remaining elements are shifted left to fill the emptied indices
  
  - brute-force
  
  - time O(n): a pointer to keep track of first vacant entry. iterate through A, move new value to first vacant entry, then move first vacant entry += 1

In [0]:
# returns the number of valid entries after deletion

def delete_duplicates(A):
  if not A:
    return 0
  
  write_index = 1
  for i in range(1, len(A)):
    if A[write_index-1]!=A[i]:
      A[write_index] = A[i] # record the A[pointer]
      write_index += 1 # move pointer
  return write_index
  
 

delete_duplicates([2,3,5,5,7,11,11,11,13])

6

## 5.6 buy and sell a stock once

  input: an array of stock prices
  
  output: max profit after buying one and then selling it later
  
  - focus on valid differences: compute the difference of the current entry with the minimum value seen so far as we iterate through the array
    - time O(n), space O(1) where n is len(A)

In [0]:
import sys
import random

def buy_and_sell_stock_once(prices):
  min_price_so_far, max_profit = float('inf'), 0.0
  for price in prices:
    max_profit_sell_today = price - min_price_so_far
    max_profit = max(max_profit, max_profit_sell_today)
    min_price_so_far = min(min_price_so_far, price)
  return max_profit


def check_ans(h):
    """O(n^2) checking answer."""
    return max([h[i] - h[j] for i in range(1, len(h)) for j in range(i)],
               default=0)


def main():
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 1000)
        a = [random.uniform(0, n) for _ in range(n)]
        #print(buy_and_sell_stock_once(a))
        assert check_ans(a) == buy_and_sell_stock_once(a)


if __name__ == '__main__':
    main()

## 5.7 buy and sell a stock twice

  input: an array of stock prices
  
  output: max profit after buyng and selling a share at most twice
  
  
    - brute-force: examine all combination of buy-sell-buy-sell with O(n^4) time complexity
    
    - two phases: F[0,j]+B[j,n-1]
      - A[0,j]: most profit by day i (record min price)
      - A[j,n-1]: most profit on and after day i (record max price)
     - M[i]=F[i-1]+F[i]
      

In [0]:
import math
import sys
import random

def buy_and_sell_stock_twice(prices):
  max_total_profit, min_price_so_far = 0.0, float('inf')
  first_buy_sell_profits = [0]* len(prices)
  # forward phase: for each day, we record max profit if we sell on that day
  for i, price in enumerate(prices):
    min_price_so_far = min(min_price_so_far, price)
    max_total_profit = max(max_total_profit, price- min_price_so_far)
    first_buy_sell_profits[i] = max_total_profit
    
  # backward phase: for each day, we record the max profit if we make second buy on that day
  max_price_so_far = float('-inf')
  for i, price in reversed(list(enumerate(prices[1:],1))):
    max_price_so_far = max(max_price_so_far, price)
    max_total_profit = max(max_total_profit, max_price_so_far - price + first_buy_sell_profits[i-1])
  return max_total_profit


def buy_and_sell_stock_twice_constant_space(prices):
    min_prices, max_profits = [float('inf')] * 2, [0] * 2
    for price in prices:
        for i in reversed(list(range(2))):
            max_profits[i] = max(max_profits[i], price - min_prices[i])
            min_prices[i] = min(min_prices[i],
                                price - (0 if i == 0 else max_profits[i - 1]))
    return max_profits[-1]


def main():
    for _ in range(1000):
        n = int(sys.argv[1]) if len(sys.argv) == 2 else random.randint(1, 100)
        a = [random.uniform(0, 10000) for _ in range(n)]
        assert math.isclose(
            buy_and_sell_stock_twice_constant_space(a),
            buy_and_sell_stock_twice(a))


if __name__ == '__main__':
    main()

In [0]:
list(enumerate([3,1,4],1))

[(1, 3), (2, 1), (3, 4)]

In [0]:
list(reversed(list(enumerate([3,1,4],1))))

[(3, 4), (2, 1), (1, 3)]

## best time to buy and sell stock (121)

  input: an stock price array
  need: one transaction (buy one and sell one)
  output: most profit
  
  - record the max cash after buying stock at any point previously (which is negative), and the max cash after selling as the cash after buying and selling.
  - time O(n), space O(1)

In [0]:
def maxProfit(prices):
        
        buy = float('-inf') # max cash balance after buying a stock
        sell = 0 # max cash balance afte rbuying and selling a stock
        
        for price in prices:
            buy = max(-price, buy) # max of buying earlier or now
            sell = max(price+buy, sell) # max of selling earlier or now
            
        return sell

## best time to buy and sell sstock II (122)

  input: an stock price array
  rule: complete as many trasactions as you like, must sell the stock before you buy again
  output: max profit
  
  - sum all price increases
  - time O(n), space O(n) could be O(1) without creating a new list

In [0]:
def maxProfit(prices):
  return sum([max(prices[i]-prices[i-1],0) for i in range(1,len(prices))])

## best time to buy and sell stock III (123)

  input: a stock price array
  rule: you may complete at most two transactions, you must sell the stock before you buy again
  output: max profit
  
  - As per problem 121 with an additional buy and sell
  - time O(n), space O(1)
  
 

In [0]:
def maxProfit(prices):
  buy1, buy2 = float('-inf'),float('-inf')
  sell1, sell2 = 0,0
  
  for price in prices:
    buy1 = max(buy1, -price)
    sell1 = max(sell1, buy1 + price)
    buy2 = max(buy2, sell1-price)
    sell2 = max (sell2, buy2 + price)
  return sell2

## best time to buy and sell stock with cooldown (309)

  input: price stock list
  rule: as many transactions as you like, must sell the stock before you buy again. after sell, you can't buy stock on next day, i.e. cooldown  1 day
  
  - dynamic programming

## enumerate

   enumerate(iterable, start=0)
  
  - Iterable: any object that supports iteration
  - Start: the index value from which the counter is to be started, by default it is 0 

In [0]:
# Python program to illustrate 
# enumerate function 
l1 = ["eat","sleep","repeat"] 
s1 = "geek"
  
# creating enumerate objects 
obj1 = enumerate(l1) 
obj2 = enumerate(s1) 
  
print ("Return type:",type(obj1) )
print (list(enumerate(l1)) )
  
# changing start index to 2 from 0 
print (list(enumerate(s1,2)) )

Return type: <class 'enumerate'>
[(0, 'eat'), (1, 'sleep'), (2, 'repeat')]
[(2, 'g'), (3, 'e'), (4, 'e'), (5, 'k')]


In [3]:
[[0 for i in range(2)] for j in range(3)]

[[0, 0], [0, 0], [0, 0]]

In [4]:
memo={}
memo[(1,1)]=2
memo[(1,0)]=1
memo[(0,1)]=3
memo

{(0, 1): 3, (1, 0): 1, (1, 1): 2}